File: backend/app/core/network_policy.py
from __future__ import annotations

from dataclasses import dataclass
from ipaddress import ip_address

from .config import Settings


@dataclass(frozen=True)
class BindHosts:
    backend: str
    frontend: str
    lan_enabled: bool


def is_loopback_host(host: str) -> bool:
    normalized = host.strip().lower().strip("[]")
    if normalized == "localhost":
        return True
    try:
        return ip_address(normalized).is_loopback
    except ValueError:
        return False


def resolve_bind_hosts(settings: Settings) -> BindHosts:
    backend = settings.backend_host.strip()
    frontend = settings.frontend_host.strip()
    if not backend or not frontend:
        raise ValueError("BACKEND_HOST and FRONTEND_HOST must not be empty")
    if not settings.allow_lan_access:
        unsafe = [host for host in (backend, frontend) if not is_loopback_host(host)]
        if unsafe:
            raise ValueError(
                "Non-loopback binding requires ALLOW_LAN_ACCESS=true: "
                + ", ".join(unsafe)
            )
    return BindHosts(
        backend=backend,
        frontend=frontend,
        lan_enabled=settings.allow_lan_access,
    )

File: frontend/src/config/networkPolicy.ts
const LOOPBACK_HOST = "127.0.0.1";

interface NetworkEnvironment {
  ALLOW_LAN_ACCESS?: string;
  BACKEND_HOST?: string;
  FRONTEND_HOST?: string;
}

export interface FrontendNetworkPolicy {
  backendProxyHost: string;
  frontendHost: string;
  lanEnabled: boolean;
}

function normalizedHost(host: string): string {
  return host
    .trim()
    .toLowerCase()
    .replace(/^\[|\]$/g, "");
}

function isValidLoopbackIpv4(host: string): boolean {
  const octets = host.split(".");
  return (
    octets.length === 4 &&
    octets[0] === "127" &&
    octets.every((octet) => {
      if (!/^\d+$/.test(octet)) return false;
      const value = Number(octet);
      return value >= 0 && value <= 255;
    })
  );
}

export function isLoopbackHost(host: string): boolean {
  const normalized = normalizedHost(host);
  return (
    normalized === "localhost" ||
    normalized === "::1" ||
    normalized === "0:0:0:0:0:0:0:1" ||
    isValidLoopbackIpv4(normalized)
  );
}

function readHost(value: string | undefined, name: string): string {
  if (value === undefined) return LOOPBACK_HOST;
  const host = value.trim();
  if (!host) throw new Error(`${name} must not be empty`);
  return host;
}

function proxyHostFor(backendHost: string): string {
  const normalized = normalizedHost(backendHost);
  if (normalized === "0.0.0.0" || normalized === "::") {
    return LOOPBACK_HOST;
  }
  return normalized.includes(":") ? `[${normalized}]` : backendHost;
}

export function resolveFrontendNetworkPolicy(env: NetworkEnvironment): FrontendNetworkPolicy {
  const backendHost = readHost(env.BACKEND_HOST, "BACKEND_HOST");
  const frontendHost = readHost(env.FRONTEND_HOST, "FRONTEND_HOST");
  const lanEnabled = env.ALLOW_LAN_ACCESS?.trim().toLowerCase() === "true";

  if (!lanEnabled) {
    const unsafe = [backendHost, frontendHost].filter((host) => !isLoopbackHost(host));
    if (unsafe.length > 0) {
      throw new Error(`Non-loopback binding requires ALLOW_LAN_ACCESS=true: ${unsafe.join(", ")}`);
    }
  }

  return {
    backendProxyHost: proxyHostFor(backendHost),
    frontendHost,
    lanEnabled,
  };
}
